prov/verbs/src/fi_verbs.h

```diff
@@ -354,6 +362,8 @@ struct fi_ibv_domain {
 	fi_ibv_mr_reg_cb		internal_mr_reg;
 	fi_ibv_mr_dereg_cb		internal_mr_dereg;
 	struct fi_ibv_mem_notifier	*notifier;
+	fi_ibv_post_send_func		post_send;
+	fi_ibv_poll_cq_func		poll_cq;
```

Comment:
You don't need to create typedefs.  You can just define the function pointers as part of struct fi_ibv_domain  This is probably true for the mr callbacks as well.

Also, you may be able to set post_send=ibv_post_send(), and poll_cq=ibv_poll_cq() and avoid the wrappers when calling verbs directly.  To do this you would need pass in the ibv_qp and ibv_cq directly to the calls, and be able to use the ibv_qp->context and ibv_cq->context to get back to your structures for the tracking versions.  Hopefully that makes sense.

In fact, if you go further, you might be able to set the function pointers to call the ibv verbs provider post/poll calls directly and avoid the ibv_post/poll wrappers.

Reply:
>You don't need to create typedefs. You can just define the function pointers as part of struct fi_ibv_domain This is probably true for the mr callbacks as well.
Also, you may be able to set post_send=ibv_post_send(), and poll_cq=ibv_poll_cq() and avoid the wrappers when calling verbs directly. To do this you would need pass in the ibv_qp and ibv_cq directly to the calls, and be able to use the ibv_qp->context and ibv_cq->context to get back to your structures for the tracking versions. Hopefully that makes sense.

👍 
>In fact, if you go further, you might be able to set the function pointers to call the ibv verbs provider post/poll calls directly and avoid the ibv_post/poll wrappers.

We would still need the wrappers that check / update the credits, right?

Reply:
Yes, the wrappers for the credits are still needed.  Those would basically look like what you have already.